### PR TITLE
[WIP] Add Base64 serialization for cursors (#31)

### PR DESCRIPTION
**Pull request by FabienChaynes**
This PR adds a new `Mongoid::Scroll::Base64EncodedCursor` class inheriting `Mongoid::Scroll::Cursor` but overriding ` Mongoid::Scroll::Base64EncodedCursor#to_s` to serialize the cursor using Base64 (cf. #31).

The `Mongoid::Scroll::Base64EncodedCursor.deserialize` class method allows to get a new `Mongoid::Scroll::Base64EncodedCursor` instance from a Base64 string.

The `Mongoid::Scroll::Base64EncodedCursor.from_cursor` class method allows to get a new `Mongoid::Scroll::Base64EncodedCursor` instance from a `Mongoid::Scroll::Cursor` instance.

This was started in #28 but led to other questions, so it was split into several PRs. The current one is only about the cursors Base64 serialization.

### CODE REVIEW

**Comment by FabienChaynes** [comment]
This PR will need more tests, README updates, etc...

I'll handle it if we decide to go further with this approach.

**Review by dblock — CHANGES_REQUESTED** [review]
LGTM, I'm interested in how this will be used.

Naming-wise I think deserialize should be called `parse` and go into the class.

I don't think you need `from_cursor` at all. If one decides to use `Base64EncodedCursor` it's already a cursor. What's the scenario where a regular cursor needs to mutate into this one?

**Comment by FabienChaynes** [comment]
> Naming-wise I think deserialize should be called parse and go into the class.

Done in 065dbfb97d3d3268756dd55f42e9b223c64589a2.

> I don't think you need from_cursor at all. If one decides to use Base64EncodedCursor it's already a cursor. What's the scenario where a regular cursor needs to mutate into this one?

Here's my usecase:
I'm building an API, I'm using `scroll` and I want to return in my API a Base64 encoded cursor to allow the API user to make a call requiring the next page.
`from_record` returns a `Cursor` but I need a `Base64EncodedCursor`:
```ruby\
cursor = params["cursor"] ? Mongoid::Scroll::Base64EncodedCursor.parse(params["cursor"]) : nil
records = []
next_page_cursor = nil

criteria.scroll(cursor) do |record, next_cursor|
  next_page_cursor = next_cursor
  records << record
end

return {
  records: records,
  paging: {
    next_cursor: Mongoid::Scroll::Base64EncodedCursor.from_cursor(next_cursor).to_s
  }
}
```

Then the API user will be able to call the API with something like `GET https://api.example.com/items?cursor=eyJ2YWx1ZSI6bnVsbCwiZmllbGRfdHlwZSI6IlN0cmluZyIsImZpZWxkX25hbWUiOiJhX3N0cmluZyIsImRpcmVjdGlvbiI6MX0=` with the value of the `cursor` URL param being the one from `paging['next_cursor']` in the response body.

File: lib/mongoid/scroll/base64_encoded_cursor.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+      def to_s
+        Base64.strict_encode64({ value: super, field_type: field_type, field_name: field_name, direction: direction }.to_json)
+      end
+
+      class << self
+        def deserialize(str)
+          config_hash = ::JSON.parse(::Base64.strict_decode64(str))
+          new(config_hash['value'], field_type: config_hash['field_type'], field_name: config_hash['field_name'], direction: config_hash['direction'])
+        rescue ArgumentError
+          raise Mongoid::Scroll::Errors::InvalidBase64Error.new(str: str)
```
[the remark this answers is not in the document]
Comment (FabienChaynes) [reply lib/mongoid/scroll/base64_encoded_cursor.rb:17]: My goal was mainly to not raise the error directly from `Base64.strict_decode64` if the Base64 string was invalid:
```ruby
[1] pry(main)> Base64.strict_decode64('foo')
ArgumentError: invalid base64
from <internal:pack>:281:in `unpack1'
```

But here it also wraps `new` which can also raise an `ArgumentError` from `Mongoid::Scroll::Cursor.extract_field_options`.

I don't really know what's best:

1. Wrapping only `Base64.strict_decode64` to have a custom exception instead of the original one
2. Wrapping everything, in which case an `InvalidBase64CursorError` probably makes more sense
3. Not rescue at all, and let the original exception (from `Base64`, `JSON`, `Mongoid::Scroll::Cursor` or anything else) bubble up

I think I'd go with 3. and just remove the rescue/not create a custom exception.

File: lib/mongoid/scroll/base64_encoded_cursor.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+module Mongoid
+  module Scroll
+    # Allows to serializer/deserialize the cursor using RFC 4648
+    class Base64EncodedCursor < Cursor
+      def to_s
+        Base64.strict_encode64({ value: super, field_type: field_type, field_name: field_name, direction: direction }.to_json)
+      end
+
+      class << self
+        def parse(str)
```
Comment (dblock) [inline lib/mongoid/scroll/base64_encoded_cursor.rb:13]: In `Cursor` [this is an instance method](https://github.com/mongoid/mongoid-scroll/blob/master/lib/mongoid/scroll/cursor.rb#LL10C21-L10C21), and a cursor can be initialize from a value. Override that method.

```
class Base64EncodedCursor < Cursor
  def parse(value)
    # base64 implementation
  end
end
```

With this `Base64EncodedCursor` will work like `Cursor`, aka `Base64EncodedCursor.new(value, options ...)`.

File: lib/mongoid/scroll/base64_encoded_cursor.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (dblock) [reply lib/mongoid/scroll/base64_encoded_cursor.rb:17]: Understood. I think you should raise an `InvalidCursorError` just like [here](https://github.com/mongoid/mongoid-scroll/blob/master/lib/mongoid/scroll/cursor.rb#L71). If there's additional information and you want to include something specific about the base64 error, subclass `InvalidCursorError` into `InvalidBase64CursorError` (this way anyone can rescue any `InvalidCursorError` like before). I was mostly commenting on the naming.

**Review by dblock — CHANGES_REQUESTED** [review]
See below.

File: lib/mongoid/scroll/base64_encoded_cursor.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (FabienChaynes) [reply lib/mongoid/scroll/base64_encoded_cursor.rb]: Done in 2415bb3bc944f6d0aa354bed5955c82028cd90d1.

File: lib/mongoid/scroll/base64_encoded_cursor.rb (hunk 2 above)
Comment (FabienChaynes) [reply lib/mongoid/scroll/base64_encoded_cursor.rb]: I tried something in 93185a37ac24780a45e2f7eb31f8b056f76687c8.

Basically what is done in `initialize` in `Cursor` doesn't make sense for `Base64EncodedCursor`. With a `Base64EncodedCursor`, what is set int the `options` hash is useless because everything is contained in the Base64 string. So every attribute that is set there will be overridden in `Base64EncodedCursor#parse`. I also had to update `Cursor.extract_field_options` to not raise when it's called for a `Base64EncodedCursor` and the `field` options are missing (since it's not required anymore).

Another way could be to just override `initialize` entirely in `Base64EncodedCursor` and do what is currently done in `Base64EncodedCursor#parse`. But it might be detrimental if we think that later on we'll have common parts in the initializers between all cursors.

File: lib/mongoid/scroll/base64_encoded_cursor.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
+require 'base64'
+require 'json'
+
+module Mongoid
+  module Scroll
+    # Allows to serializer/deserialize the cursor using RFC 4648
+    class Base64EncodedCursor < Cursor
+      class << self
+        def from_cursor(cursor)
```
Comment (dblock) [inline lib/mongoid/scroll/base64_encoded_cursor.rb:9]: What's a valid use-case for mutating a cursor into a base64 cursor? If there isn't one, I think this shouldn't exist.

File: lib/mongoid/scroll/base64_encoded_cursor.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+                                      direction: cursor.direction,
+                                      include_current: cursor.include_current)
+          base64_encoded_cursor.value = cursor.value
+          base64_encoded_cursor.tiebreak_id = cursor.tiebreak_id
+          base64_encoded_cursor
+        end
+      end
+
+      def to_s
+        Base64.strict_encode64({ value: super, field_type: field_type, field_name: field_name, direction: direction, include_current: include_current }.to_json)
```
Comment (dblock) [inline lib/mongoid/scroll/base64_encoded_cursor.rb:22]: Nit: it's a long line, maybe put each value on its own - it's also easier to read diffs.

```
({
  value: super,
  field_type: ...
   ...
})
```

File: lib/mongoid/scroll/base64_encoded_cursor.rb [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        begin
+          config_hash = ::JSON.parse(::Base64.strict_decode64(value))
+        rescue
+          raise Mongoid::Scroll::Errors::InvalidCursorError.new(cursor: value)
+        end
+        @field_type = config_hash['field_type']
+        @field_name = config_hash['field_name']
+        @direction = config_hash['direction']
+        @include_current = config_hash['include_current']
+        super(config_hash['value'])
```
Comment (dblock) [inline lib/mongoid/scroll/base64_encoded_cursor.rb:39]: Pass field options down, this way you can remove the check for class in `extract_field_options`.

```
super config_hash['value'], config_hash
```

File: lib/mongoid/scroll/cursor.rb [hunk 6]
```diff
@@ -86,7 +86,7 @@ def extract_field_options(options)
             [field_type.to_s, field_name.to_s]
           elsif options && (field = options[:field])
             [field.type.to_s, field.name.to_s]
-          else
+          elsif self == Mongoid::Scroll::Cursor
```
Comment (dblock) [inline lib/mongoid/scroll/cursor.rb:89]: Undo this change, see above.

**Review by dblock — CHANGES_REQUESTED** [review]
Almost there! Needs README update.

File: lib/mongoid/scroll/base64_encoded_cursor.rb (hunk 2 above)
Comment (dblock) [reply lib/mongoid/scroll/base64_encoded_cursor.rb]: The third option is to pass options down, which avoids the base class change, I added a comment below.

File: lib/mongoid/scroll/base64_encoded_cursor.rb (hunk 5 above)
Comment (FabienChaynes) [reply lib/mongoid/scroll/base64_encoded_cursor.rb:39]: `super` is calling `Cursor#parse` which only accepts a `value` argument, not an extra `options` one. It's not this method which is raising (because of `Cursor.extract_field_options`) if the `:field_name`/`:field_name` options are missing, it's the `initialize` method.

If we always raise in `Cursor.extract_field_options` (removing the `elsif self == Mongoid::Scroll::Cursor` that I added), we won't be able to create a `Base64EncodedCursor` without option (`Mongoid::Scroll::Base64EncodedCursor.new(base64_string)` will raise).

File: lib/mongoid/scroll/base64_encoded_cursor.rb (hunk 3 above)
Comment (FabienChaynes) [reply lib/mongoid/scroll/base64_encoded_cursor.rb:9]: The one I described in https://github.com/mongoid/mongoid-scroll/pull/32#issuecomment-1450644272.

The `scroll` methods are yielding a `Cursor`, but I need to call `Base64EncodedCursor#to_s` to return it to my API users so that they can forward it in a next request an I can use it to return the next page.

Comment (dblock) [reply lib/mongoid/scroll/base64_encoded_cursor.rb:9]: Right, I think this is not ideal from the user POV. Can we have `Feed::Item.desc(:position).limit(5).scroll(klass: Base64EncodedCursor)` or something like that instead?

File: lib/mongoid/scroll/base64_encoded_cursor.rb (hunk 5 above)
Comment (dblock) [reply lib/mongoid/scroll/base64_encoded_cursor.rb:45]: Ok I see. I think we're trying to reuse too much of the base cursor which attempts to join things with `:` and what not, all while it is completely unnecessary for a base64 cursor which is happy to just base64(hash). How about we don't call super in parse at all as you suggested before? Override the constructor and to_s?

Maybe we should have a `Base64Cursor`, a `Cursor` (string cursor) and a `Base` for common parts?

File: lib/mongoid/scroll/base64_encoded_cursor.rb (hunk 3 above)
Comment (FabienChaynes) [reply lib/mongoid/scroll/base64_encoded_cursor.rb:9]: It's what I did in 73bb69449f6745b517087ace46e5f8351212ab2c (explanation in https://github.com/mongoid/mongoid-scroll/pull/28#issuecomment-1440807583 3.a.).

We can't have `scroll(klass: Base64EncodedCursor)` because the signature is `scroll(cursor, options)` so it'll have to be in the second parameter (which means that even if `cursor` is `nil`, we'd have to pass it).
It brings a lot of changes and also questions (what do we do if a cursor is specified but is a different class from `klass`?).
The current PR is already getting bigger because of https://github.com/mongoid/mongoid-scroll/pull/32#discussion_r1125675319, so IMO I should just remove `Base64EncodedCursor.from_cursor` for now if you don't like it, focus on the other discussion (not having `Base64EncodedCursor` inherit from `Cursor`), and create another PR later on to tackle the yielded cursors once this one will be done. Good for you?

**Comment by dblock** [comment]
I'm going to merge this and try to improve upon it.
